sync.py

```diff
@@ -71,22 +82,64 @@ def all_messages(credentials, full_sync=False) -> int:
         )
 
         messages = results.get("messages", [])
-
-        total_messages += len(messages)
-        for i, m in enumerate(messages, start=total_messages - len(messages) + 1):
-            try:
-                raw_msg = (
-                    service.users().messages().get(userId="me", id=m["id"]).execute()
-                )
-                msg = message.Message.from_raw(raw_msg, labels)
-                db.create_message(msg)
-
-            except IntegrityError as e:
-                print(f"Could not process message {m['id']}: {str(e)}")
-            except TimeoutError as e:
-                print(f"Could not get message from Gmail {m['id']}: {str(e)}")
-
-            print(f"Synced message {msg.id} from {msg.timestamp} (Count: {i})")
+        message_ids = [m["id"] for m in messages]
+
+        # Process in batches (parallel fetch)
+        for batch_start in range(0, len(message_ids), BATCH_SIZE):
+            batch_ids = message_ids[batch_start:batch_start+BATCH_SIZE]
+            print(f"\nProcessing batch {batch_start // BATCH_SIZE + 1} (messages {batch_start + 1}-{batch_start + len(batch_ids)})")
+            print(f"Batch message IDs: {batch_ids}")
+
+            def fetch_message(msg_id):
+                for attempt in range(1, MAX_FETCH_RETRIES + 1):
+                    try:
+                        # Rebuild the service for each thread to avoid sharing connections
+                        local_service = build("gmail", "v1", credentials=credentials, cache_discovery=False)
```

Comment:
Would it make sense to use a pool of already created services? This would eliminate rebuilding local services. The pool could be as large as the batch size.